packages/cache/src/internal/cacheHttpClient.ts

```diff
@@ -0,0 +1,339 @@
+import * as core from '@actions/core'
+import {HttpClient, HttpCodes} from '@actions/http-client'
+import {BearerCredentialHandler} from '@actions/http-client/auth'
+import {
+  IHttpClientResponse,
+  IRequestOptions,
+  ITypedResponse
+} from '@actions/http-client/interfaces'
+import * as crypto from 'crypto'
+import * as fs from 'fs'
+import * as stream from 'stream'
+import * as util from 'util'
+
+import * as utils from './cacheUtils'
+import {CompressionMethod, SocketTimeout} from './constants'
+import {
+  ArtifactCacheEntry,
+  CacheOptions,
+  CommitCacheRequest,
+  ReserveCacheRequest,
+  ReserveCacheResponse
+} from './contracts'
+
+const versionSalt = '1.0'
+
+function isSuccessStatusCode(statusCode?: number): boolean {
+  if (!statusCode) {
+    return false
+  }
+  return statusCode >= 200 && statusCode < 300
+}
+
+function isRetryableStatusCode(statusCode?: number): boolean {
+  if (!statusCode) {
+    return false
+  }
+  const retryableStatusCodes = [
+    HttpCodes.BadGateway,
+    HttpCodes.ServiceUnavailable,
+    HttpCodes.GatewayTimeout
+  ]
+  return retryableStatusCodes.includes(statusCode)
+}
+
+function getCacheApiUrl(resource: string): string {
+  // Ideally we just use ACTIONS_CACHE_URL
+  const baseUrl: string = (
+    process.env['ACTIONS_CACHE_URL'] ||
+    process.env['ACTIONS_RUNTIME_URL'] ||
+    ''
+  ).replace('pipelines', 'artifactcache')
+  if (!baseUrl) {
+    throw new Error('Cache Service Url not found, unable to restore cache.')
+  }
+
+  const url = `${baseUrl}_apis/artifactcache/${resource}`
+  core.debug(`Resource Url: ${url}`)
+  return url
+}
+
+function createAcceptHeader(type: string, apiVersion: string): string {
+  return `${type};api-version=${apiVersion}`
+}
+
+function getRequestOptions(): IRequestOptions {
+  const requestOptions: IRequestOptions = {
+    headers: {
+      Accept: createAcceptHeader('application/json', '6.0-preview.1')
+    }
+  }
+
+  return requestOptions
+}
+
+function createHttpClient(): HttpClient {
+  const token = process.env['ACTIONS_RUNTIME_TOKEN'] || ''
+  const bearerCredentialHandler = new BearerCredentialHandler(token)
+
+  return new HttpClient(
+    'actions/cache',
+    [bearerCredentialHandler],
+    getRequestOptions()
+  )
+}
+
+export function getCacheVersion(
+  inputPath: string,
```

Comment:
`inputPath` here is the `path` input variable for our first party cache action. We use the input string directly to compute version [here](https://github.com/actions/cache/blob/ce9276c90e52c28f187062530b2d7a43fd7973a5/src/cacheHttpClient.ts#L89). 

This is not ideal for the cache package. Action authors might want to implement their cache action differently and use different variables. Perhaps we should pass paths as an array to `getCacheVersion` and bump up the version salt.

Reply:
I think it would make sense to allow action authors to optionally specify a `cacheVersion` along with their key and we could use this function to generate one if not specified 

> I would like to change this to pass paths as an array to getCacheVersion but it would break existing caches.

There are no consumers of this package yet, so we can make that breaking change. It will be breaking when we start consuming it in `actions/cache` but that's acceptable since this will go along with the breaking v2 major version.